File: Algorithm/leetcode_046.go
/**
 * File              : leetcode_046.go
 * Author            : Andy
 * Date              : 2022.08.15
 * Last Modified Date: 2022.08.15
 * Last Modified By  : Andy
 */


func permute(nums []int) [][]int {
    var result [][]int
    var path []int
    used := make([]bool, len(nums))
    find(nums, len(nums), path, used, &result)
    return result
}


func (nums []int, n int, path []int, used []bool, result *[][]int){
    // 终止条件
    if len(path) == n {
        p := make([]int, len(path))
        copy(p, path)
        *result = append(*result, p)
        return
    }

    for i:=0;i<n;i++{
        if used[i] == true{
            continue
        }
        path = append(path, nums[i])
        used[i] = true
        find(nums, n, path, used, result)
        // 还原
        path = path[:len(path)-1]
        used[i] = false
    }
}

File: Algorithm/leetcode_046.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# File              : leetcode_046.py
# Author            : Andy
# Date              : 2022.08.15
# Last Modified Date: 2022.08.15
# Last Modified By  : Andy


class Solution:
    def permute(self, nums):
        n = len(nums)
        path = []
        result = []

        def find(index):
            if index == n:
                result.append(path[:])
                return

            for i in range(n):
                if nums[i] in path:
                    continue
                path.append(nums[i])
                find(index+1)
                path.pop()
        find(0)
        return result
